### DISCUSSION TITLE
Recipes for Pre-release versions without initial breaking changes

### DISCUSSION THREAD

**Discussion by nitzano** [semantic-release/semantic-release#2142] [Q&A] [answered]
First thank you for this amazing library.

I've looked in the [Publishing pre-releases](https://github.com/semantic-release/semantic-release/blob/master/docs/recipes/pre-releases.md) recipe and couldn't find cases where the pre-release (`beta` for example) does not start with a breaking changes (which can happen sometimes).

For example I've noticed (as demostrated below in recipe#2) - that if I'm pre-releasing a `fix` and then adding `feat` commits to it, the beta version will be actually be updated to a new one, resetting the "beta counter" back to `beta.1`

Are my examples correct or am I missing something? (added `q:` with all my checks)

## Recipe-1: Pre-release of a feature with fixes 

```
* feat: initial commit # => v1.0.0 on @latest
| \
|  * feat: first feat on beta            # => v1.1.0-beta.1 on @beta
|  * feat: second feat on beta           # => v1.1.0-beta.2 on @beta
|  * fix: first fix on beta              # => v1.1.0-beta.3 on @beta
|  * feat: third feat on beta            # => v1.1.0-beta.4 on @beta
*  | fix: a fix on master # => v1.0.1 on @latest
*  |  | Merge branch beta into master # => v1.1.0 on @latest
```

## Recipe-2: Pre-release of a fix with features
```
* feat: initial commit # => v1.0.0 on @latest
| \
|  * fix first fix on beta                # => v1.0.1-beta.1 on @beta
|  * feat: first feat on beta             # => v1.1.0-beta.1 on @beta <-- q: is it expected to reset to 1.1.0-beta.1?
|  * fix: second fix on beta              # => v1.1.0-beta.2 on @beta
|  * feat: second feat on beta            # => v1.1.0-beta.3 on @beta
*  | fix: a fix on master # => v1.0.1 on @latest
*  |  | Merge branch beta into master # => v1.1.0 on @latest
```

I also have some more cases (would love to open pull request to add them if needed) I'm thinking of:

## Recipe-3: Pre-release with a conflict with the default channel
```
* feat: initial commit # => v1.0.0 on @latest
| \
|  * feat: first feat on beta            # => v1.1.0-beta.1 on @beta
|  * feat: second feat on beta           # => v1.1.0-beta.2 on @beta
|  * fix: first fix on beta              # => v1.1.0-beta.3 on @beta
|  * feat: third feat on beta            # => v1.1.0-beta.4 on @beta
*  | fix: a fix on master # => v1.0.1 on @latest
*  | feat : a feat on master # => v1.1.0 on @latest  <-- q: conflict or will it work?
*  |  | Merge branch beta into master # => v1.1.0 on @latest     
```

## Recipe-4: Pre-release of a feature with a breaking change in the middle

```
* feat: initial commit # => v1.0.0 on @latest
| \
|  * feat: first feat on beta            # => v1.1.0-beta.1 on @beta
|  * feat: second feat on beta           # => v1.1.0-beta.2 on @beta
|  * fix: first fix on beta              # => v1.1.0-beta.3 on @beta
|  * feat: another feat \n\n BREAKING CHANGE: it breaks something # => v2.0.0-beta.1 on @beta 
|  * feat: second feat on beta            # => v2.1.0-beta.1 on @beta <-- q: is it correct?
*  | fix: a fix on master # => v1.0.1 on @latest
*  |  | Merge branch beta into master # => v2.1.0 on @latest 
```

**Comment by travi** [comment, marked as answer]
you have recipe 1 and recipe 2 correct.

### Recipe 3

the conflict isnt so much a conflict as is it an example where the actual predicted version of a pre-release is just a prediction. it does not necessarily mean that it will release that number once promoted to stable. instead, it communicates the impact of the chance since the last version that is known based on the history of that branch. if additional changes exist in the mainline branch, the same impact will be taken into account when calculating the proper version bump, but from the adjusted starting point.

to make that concrete, there are two options for you that would work out with your example:

merge to the mainline and let the starting point be sorted out at that point

```
* feat: initial commit # => v1.0.0 on @latest
| \
|  * feat: first feat on beta            # => v1.1.0-beta.1 on @beta
|  * feat: second feat on beta           # => v1.1.0-beta.2 on @beta
|  * fix: first fix on beta              # => v1.1.0-beta.3 on @beta
|  * feat: third feat on beta            # => v1.1.0-beta.4 on @beta
*  | fix: a fix on master # => v1.0.1 on @latest
*  | feat : a feat on master # => v1.1.0 on @latest
*  |  | Merge branch beta into master # => v1.2.0 on @latest     
```

or, update the history of the beta branch so that it is aware of the new changes on the mainline branch

```
* feat: initial commit # => v1.0.0 on @latest
| \
|  * feat: first feat on beta            # => v1.1.0-beta.1 on @beta
|  * feat: second feat on beta           # => v1.1.0-beta.2 on @beta
|  * fix: first fix on beta              # => v1.1.0-beta.3 on @beta
|  * feat: third feat on beta            # => v1.1.0-beta.4 on @beta
*  | fix: a fix on master # => v1.0.1 on @latest
*  | feat : a feat on master # => v1.1.0 on @latest
*  |  | Merge branch master into beta # => v1.2.0-beta.1 on @beta     
*  |  | Merge branch beta into master # => v1.2.0 on @latest     
```

### Recipe 4

```
* feat: initial commit # => v1.0.0 on @latest
| \
|  * feat: first feat on beta            # => v1.1.0-beta.1 on @beta
|  * feat: second feat on beta           # => v1.1.0-beta.2 on @beta
|  * fix: first fix on beta              # => v1.1.0-beta.3 on @beta
|  * feat: another feat \n\n BREAKING CHANGE: it breaks something # => v2.0.0-beta.1 on @beta 
|  * feat: second feat on beta            # => v2.0.0-beta.2 on @beta <-- a: the breaking change is higher impact than the fix, so no further bump of the predicted version
*  | fix: a fix on master # => v1.0.1 on @latest
*  |  | Merge branch beta into master # => v2.0.0 on @latest 
```

**Reply by nitzano** [reply]
@travi Thank you for the quick reply!

**Reply by travi** [reply]
sure thing. if you think that the docs could be updated to make these situation more clear, we would be happy to consider a PR with your perspective while you still have the thoughts fresh in your mind.

**Reply by nitzano** [reply]
I think it could help to have more examples of pre-releases where the version doesn't start with breaking changes as most of them aren't (at least not in my projects) so it's a common practice, I'll open a PR so it can be reviewed.

**Reply by nitzano** [reply]
@travi: Opened https://github.com/semantic-release/semantic-release/pull/2145 for recipes #1 + #2
